File: Python/067 Add Binary.py
# Given two binary strings, return their sum (also a binary string).

# The input strings are both non-empty and contains only characters 1 or 0.

# Example 1:

# Input: a = "11", b = "1"
# Output: "100"
# Example 2:

# Input: a = "1010", b = "1011"
# Output: "10101"


class Solution(object):
    def addBinary(self, a, b):
        """
        :type a: str
        :type b: str
        :rtype: str
        """
        result = ''
        aI = len(a) - 1
        bI = len(b) - 1
        carry = 0
        while aI >= 0 or bI >= 0:
            aNum = 0
            bNum = 0
            if aI >= 0:
                aNum = int(a[aI])
            if bI >= 0:
                bNum = int(b[bI])

            num = aNum ^ bNum ^ carry
            # carry = aNum ^ bNum ? carry : aNum
            if not aNum ^ bNum:
                carry = aNum

            result = str(num) + result
            aI -= 1
            bI -= 1

        if carry > 0:
            result = str(carry) + result

        return result

# test
s = Solution()
print(s.addBinary('1', '111'))

File: JavaScript/067 Add Binary.js
// Given two binary strings, return their sum (also a binary string).

// The input strings are both non-empty and contains only characters 1 or 0.

// Example 1:

// Input: a = "11", b = "1"
// Output: "100"
// Example 2:

// Input: a = "1010", b = "1011"
// Output: "10101"


/**
 * @param {string} a
 * @param {string} b
 * @return {string}
 */
var addBinary = function (a, b) {
    result = '';
    aI = a.length - 1;
    bI = b.length - 1;
    carry = 0;
    while (aI >= 0 || bI >= 0) {
        let aNum = (aI >= 0) ? Number(a[aI]) : 0;
        let bNum = (bI >= 0) ? Number(b[bI]) : 0;
        let num = aNum ^ bNum ^ carry;
        carry = aNum ^ bNum ? carry : aNum;
        result = num.toString() + result;
        aI--;
        bI--;
    }
    if (carry > 0) result = carry.toString() + result;
    return result
};

// test
console.log(addBinary('1', '111'));